File: app/src/main/java/be/hcpl/android/filmtag/adapter/FrameAdapter.java
package be.hcpl.android.filmtag.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import be.hcpl.android.filmtag.R;
import be.hcpl.android.filmtag.model.Frame;
import be.hcpl.android.filmtag.util.TextUtil;

public class FrameAdapter extends BaseAdapter {

    private List<Frame> items = new ArrayList<>();

    private LayoutInflater mInflater;

    public FrameAdapter(final Context context) {
        mInflater = LayoutInflater.from(context);
    }

    @Override
    public int getCount() {
        return items.size();
    }

    @Override
    public Frame getItem(int i) {
        return items.get(i);
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public View getView(
            int i,
            View convertView,
            ViewGroup viewGroup) {

        final Frame frame = items.get(i);

        // no need to continue if we have no data
        if (frame == null) {
            return convertView;
        }

        View rowView = convertView;
        if (rowView == null) {
            rowView = mInflater.inflate(R.layout.list_item_frame, viewGroup, false);

            final ViewHolder holder = new ViewHolder();
            holder.textFrame = (TextView) rowView.findViewById(R.id.text_frame);
            holder.textApertureAndShutter = (TextView) rowView.findViewById(R.id.text_aperture_and_shutter);
            holder.textNotes = (TextView) rowView.findViewById(R.id.text_notes);
            rowView.setTag(holder);
        }
        final ViewHolder holder = (ViewHolder) rowView.getTag();

        // First line: frame number, aperture, shutter speed
        holder.textFrame.setText(TextUtil.formatFrameNumber(frame.getNumber()));
        holder.textApertureAndShutter.setText(formatApertureAndShutter(
                frame.getAperture(),
                frame.getShutter(),
                frame.isLongExposure()));
        // Second line: notes
        holder.textNotes.setText(frame.getNotes());

        return rowView;
    }

    private String formatApertureAndShutter(double aperture, int shutter, boolean longExposure) {
        String str = "";
        if (aperture != Frame.EMPTY_VALUE) {
            str += TextUtil.formatAperture(aperture);
        }
        if (aperture != Frame.EMPTY_VALUE && shutter != Frame.EMPTY_VALUE) {
            str += " - ";
        }
        if (shutter != Frame.EMPTY_VALUE) {
            str += TextUtil.formatShutter(shutter, longExposure);
        }
        return str;
    }

    public void clear() {
        items.clear();
    }

    public void addAll(List<Frame> frames) {
        items.addAll(frames);
    }

    class ViewHolder {

        TextView textApertureAndShutter, textFrame, textNotes;
    }
}

File: app/src/main/java/be/hcpl/android/filmtag/adapter/FrameAdapter.kt
package be.hcpl.android.filmtag.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import be.hcpl.android.filmtag.EditFrameFragment

import java.util.ArrayList

import be.hcpl.android.filmtag.R
import be.hcpl.android.filmtag.model.Frame
import be.hcpl.android.filmtag.util.TextUtil

class FrameAdapter(context: Context) : BaseAdapter() {

    private val items = ArrayList<Frame>()

    private val mInflater: LayoutInflater

    init {
        mInflater = LayoutInflater.from(context)
    }

    override fun getCount(): Int {
        return items.size
    }

    override fun getItem(i: Int): Frame {
        return items[i]
    }

    override fun getItemId(i: Int): Long {
        return i.toLong()
    }

    override fun getView(
            i: Int,
            convertView: View?,
            viewGroup: ViewGroup): View? {

        val frame = items[i] ?: return convertView

        // no need to continue if we have no data

        var rowView: View? = convertView
        if (rowView == null) {
            rowView = mInflater.inflate(R.layout.list_item_frame, viewGroup, false)

            val holder = ViewHolder()
            holder.textFrame = rowView!!.findViewById(R.id.text_frame) as TextView
            holder.textDate = rowView!!.findViewById(R.id.text_date) as TextView
            holder.textApertureAndShutter = rowView.findViewById(R.id.text_aperture_and_shutter) as TextView
            holder.textNotes = rowView.findViewById(R.id.text_notes) as TextView
            rowView.tag = holder
        }
        val holder = rowView.tag as ViewHolder

        // First line: frame number, aperture, shutter speed
        holder.textFrame!!.text = TextUtil.formatFrameNumber(frame.number)
        holder.textDate!!.text = frame.dateTaken?.let { EditFrameFragment.dateFormatter.format(it) }
        holder.textApertureAndShutter!!.text = formatApertureAndShutter(
                frame.aperture,
                frame.shutter,
                frame.isLongExposure)
        // Second line: notes
        holder.textNotes!!.text = frame.notes

        return rowView
    }

    private fun formatApertureAndShutter(aperture: Double, shutter: Int, longExposure: Boolean): String {
        var str = ""
        if (aperture != Frame.EMPTY_VALUE.toDouble()) {
            str += TextUtil.formatAperture(aperture)
        }
        if (aperture != Frame.EMPTY_VALUE.toDouble() && shutter != Frame.EMPTY_VALUE) {
            str += " - "
        }
        if (shutter != Frame.EMPTY_VALUE) {
            str += TextUtil.formatShutter(shutter, longExposure)
        }
        return str
    }

    fun clear() {
        items.clear()
    }

    fun addAll(frames: List<Frame>) {
        items.addAll(frames)
    }

    private inner class ViewHolder {

        internal var textApertureAndShutter: TextView? = null
        internal var textFrame: TextView? = null
        internal var textDate: TextView? = null
        internal var textNotes: TextView? = null
    }
}